packages/smooth_app/lib/background/background_task_folksonomy.dart

```diff
@@ -0,0 +1,222 @@
+import 'package:smooth_app/query/product_query.dart';
+
+class BackgroundTaskFolksonomy extends BackgroundTask {
+  BackgroundTaskFolksonomy._({
+    required super.processName,
+    required super.uniqueId,
+    required super.stamp,
+    required this.barcode,
+  });
+
+  BackgroundTaskFolksonomy.fromJson(super.json)
+    : barcode = json[_jsonTagBarcode] as String,
+      super.fromJson();
+
+  final String barcode;
+
+  static const String _jsonTagBarcode = 'barcode';
+  static const OperationType _operationType = OperationType.folksonomy;
+
+  @override
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> result = super.toJson();
+    result[_jsonTagBarcode] = barcode;
+    return result;
+  }
+
+  static Future<void> addTask(
+    final String barcode,
+    final LocalDatabase localDatabase,
+  ) async {
+    final String uniqueId = await _operationType.getNewKey(
+      localDatabase,
+      barcode: barcode,
+    );
+    final BackgroundTask task = _getNewTask(uniqueId, barcode);
+    await task.addToManager(localDatabase, queue: BackgroundTaskQueue.fast);
+  }
+
+  static BackgroundTask _getNewTask(
+    final String uniqueId,
+    final String barcode,
+  ) => BackgroundTaskFolksonomy._(
+    processName: _operationType.processName,
+    uniqueId: uniqueId,
+    stamp: ';folksonomy;$barcode',
+    barcode: barcode,
+  );
+
+  @override
+  (String, AlignmentGeometry)? getFloatingMessage(
+    AppLocalizations appLocalizations,
+  ) => (
+    appLocalizations.background_task_title_folksonomy,
+    AlignmentDirectional.bottomCenter,
+  );
+
+  @override
+  Future<void> execute(LocalDatabase localDatabase) async {
+    final DaoTransientFolksonomy daoTransientFolksonomy =
+        DaoTransientFolksonomy(localDatabase);
+    final DaoFolksonomy daoFolksonomy = DaoFolksonomy(localDatabase);
+
+    while (true) {
+      final FolksonomyOperation? operation = _getPendingOperations(
+        barcode,
+        daoTransientFolksonomy,
+      )?.firstOrNull;
+
+      if (operation == null) {
+        return;
+      }
+
+      if (!operation.isPerformed) {
+        final String bearerToken = await _getBearerToken();
+
+        switch (operation.type) {
+          case FolksonomyAction.add:
+            await _serverAdd(operation.tag, bearerToken);
+            break;
+          case FolksonomyAction.edit:
+            await _serverEdit(operation.tag, bearerToken);
+            break;
+          case FolksonomyAction.remove:
+            await _serverDelete(operation.tag, bearerToken);
+            break;
+          default:
+            throw Exception('Invalid operation type: ${operation.type}');
+        }
+
+        final List<FolksonomyOperation>? pendingOperations =
+            _getPendingOperations(barcode, daoTransientFolksonomy);
+        if (pendingOperations == null || pendingOperations.isEmpty) {
+          throw Exception(
+            'Not supposed to happen as we do not store null/empty lists.',
```

Comment:
More relevant: as we've just performed an operation, there should be at least this operation in the queue.